astroquery/utils/timer.py

```diff
@@ -0,0 +1,373 @@
+                result = self._timed_pfunc(arg)
+            except Exception as e:
+                warnings.warn(str(e), AstropyUserWarning)
+                self._cache_bad.append(arg)
+            else:
+                self._cache_good[arg] = result[0]  # Run time
+                self._cache_out[arg] = result[1]  # Function output
+
+    def time_func(self, arglist):
+        """Time the partial function for a list of single args
+        and store run time in a cache. This forms a baseline for
+        the prediction.
+
+        This also stores function outputs in `results`.
+
+        Parameters
+        ----------
+        arglist : list of numbers
+            List of input arguments to time.
+
+        """
+        if not isinstance(arglist, Iterable):
+            arglist = [arglist]
+
+        # Preserve arglist order
+        for arg in arglist:
+            self._cache_time(arg)
+
+    # FUTURE: Implement N^x * O(log(N)) fancy fitting.
+    def do_fit(self, model=None, fitter=None, power=1, min_datapoints=3):
+        """Fit a function to the lists of arguments and
+        their respective run time in the cache.
+
+        By default, this does a linear least-square fitting
+        to a straight line on run time w.r.t. argument values
+        raised to the given power, and returns the optimal
+        intercept and slope.
+
+        Parameters
+        ----------
+        model : `astropy.modeling.Model`
+            Model for the expected trend of run time (Y-axis)
+            w.r.t. :math:`\\text{arg}^{\\text{power}}` (X-axis).
+            If `None`, will use `~astropy.modeling.polynomial.Polynomial1D`
+            with ``degree=1``.
+
+        fitter : `astropy.modeling.fitting.Fitter`
+            Fitter for the given model to extract optimal coefficient values.
+            If `None`, will use `~astropy.modeling.fitting.LinearLSQFitter`.
+
+        power : int, optional
+            Power of values to fit.
+
+        min_datapoints : int, optional
+            Minimum number of data points required for fitting.
+            They can be built up with `time_func`.
+
+        Returns
+        -------
+        a : array-like
+            Fitted `~astropy.modeling.FittableModel` parameters.
+
+        Raises
+        ------
+        ValueError
+            Insufficient data points for fitting.
+
+        ModelsError
+            Invalid model or fitter.
+
+        """
+        # Reset related attributes
+        self._power = power
+        self._cache_est = OrderedDict()
+
+        x_arr = np.array(list(self._cache_good.keys()))
+        if x_arr.size < min_datapoints:
+            raise ValueError('requires {} points but has {}'.format(
+                min_datapoints, x_arr.size))
+
+        if model is None:
+            model = modeling.models.Polynomial1D(1)
+        elif not isinstance(model, modeling.core.Model):
+            raise modeling.fitting.ModelsError(
+                f'{model} is not a model.')
```

Comment:
we're legacy here in astroquery, and still need to support python2, so no f strings please 😅

Reply:
Oh dear... I copied it from astropy dev, which has dropped Python 2 so long ago. I gotta comb the code and see if there is another compat issues.

Reply:
if we add this to 0.4, then no need for python2, as I recall @keflavich has agreed that we can drop active support for any new stuff then (even if not yet cleaning up stuff). But python 3.6+ might be a bit stretching it.

Reply:
Yeah, sounds reasonable. I removed the f-strings. But does that mean I can **ignore** the PY2 job failure here? Please advise. Thanks!

Reply:
yeah, let's table this PR until next week, then I'll remove those builds ;)